QuarkusMock.installMock fails with ClassCastException in a Kotlin Gradle project

### Describe the bug

Quarkus version: `3.32.3`, `3.34.CR1`

When trying to replace an `ApplicationScoped` bean using `QuarkusMock.installMockForType(obj, Interface::class.java)` to supply a fake interface implementation in a `QuarkusTest`, the call fails with a `ClassCastException` because the captured bean type is not the parent interface, but its concrete production implementation, so a sibling test impl cannot be cast to it.

Documentation reference that led me to discover this error:
https://quarkus.io/guides/getting-started-testing#quarkus_mock

Based on the documentation I don't see any mistakes in the failing code.

--------

Reproduction is trivial: generate a quarkus project in the web generator with the following extensions: `quarkus-rest`, `kotlin`, build system: Gradle with Kotlin DSL, Java version: `17`

Add the following two files:

```kotlin
// src/main/kotlin/org/acme/service.kt
package org.acme

import jakarta.enterprise.context.ApplicationScoped

interface Interface

@ApplicationScoped
class ProdInterfaceImpl : Interface

@ApplicationScoped
class Service(
    private val iface: Interface
)
```

```kotlin
// src/test/kotlin/org/acme/ServiceTest.kt
package org.acme

import io.quarkus.test.junit.QuarkusMock
import io.quarkus.test.junit.QuarkusTest
import jakarta.inject.Inject
import org.junit.jupiter.api.Test

@QuarkusTest
class ServiceTest {

    @Inject
    lateinit var service: Service

    @Test
    fun testInterfaceReplacement() {
        QuarkusMock.installMockForType(TestInterfaceImpl(), Interface::class.java)
    }

    class TestInterfaceImpl : Interface
}
```

Running `testInterfaceReplacement` fails with the following:

```
org.acme.ProdInterfaceImpl@7c5d36c3 is not a normal scoped CDI bean, make sure the bean is a normal scope like @ApplicationScoped or @RequestScoped
java.lang.RuntimeException: org.acme.ProdInterfaceImpl@7c5d36c3 is not a normal scoped CDI bean, make sure the bean is a normal scope like @ApplicationScoped or @RequestScoped
	at io.quarkus.test.junit.MockSupport.installMock(MockSupport.java:63)
	at io.quarkus.test.junit.QuarkusMock.installMockForType(QuarkusMock.java:63)
	at org.acme.ServiceTest.testInterfaceReplacement(ServiceTest.kt:19)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:110)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at io.quarkus.test.junit.MockSupport.installMock(MockSupport.java:54)
	... 2 more
Caused by: java.lang.ClassCastException: class org.acme.ServiceTest$TestInterfaceImpl cannot be cast to class org.acme.ProdInterfaceImpl (org.acme.ServiceTest$TestInterfaceImpl and org.acme.ProdInterfaceImpl are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @44fdce3c)
	at org.acme.ProdInterfaceImpl_ClientProxy.arc$setMock(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 4 more```



### Expected behavior

I can supply a fake interface implementation to Quarkus in tests.

### Actual behavior

`QuarkusMock.installMockForType` fails with an exception

--- a/test-framework/junit-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SetMockitoMockAsBeanMockCallback.java
+++ b/test-framework/junit-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SetMockitoMockAsBeanMockCallback.java
@@ -24,7 +24,8 @@ public class SetMockitoMockAsBeanMockCallback implements QuarkusTestBeforeEachCa
             throw new RuntimeException(mocked.beanInstance
                     + " is not a normal scoped CDI bean, make sure the bean is a normal scope like @ApplicationScoped or @RequestScoped."
                     + " Alternatively you can use '@MockitoConfig(convertScopes = true)' in addition to '@InjectMock' if you would like"
-                    + " Quarkus to automatically make that conversion (you should only use this if you understand the implications).");
+                    + " Quarkus to automatically make that conversion (you should only use this if you understand the implications).",
+                    e);
         }
     }
 }

--- a/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusMock.java
+++ b/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusMock.java
@@ -2,33 +2,37 @@ package io.quarkus.test.junit;
 
 import java.lang.annotation.Annotation;
 
+import jakarta.enterprise.event.Event;
 import jakarta.enterprise.inject.spi.CDI;
 import jakarta.enterprise.util.TypeLiteral;
 
 import org.junit.jupiter.api.TestInstance;
 
+import io.quarkus.runtime.MockedThroughWrapper;
+
 /**
  * Utility class that can be used to mock CDI normal scoped beans.
- *
+ * <p>
  * This includes beans that are {@link jakarta.enterprise.context.ApplicationScoped} and
  * {@link jakarta.enterprise.context.RequestScoped}.
- *
- * To use this inject the bean into a test, and then invoke the mock
- * method with your mock.
- *
+ * <p>
+ * To use this, inject the bean into a test and invoke {@link #installMockForInstance(Object, Object, Options)}.
+ * Alternatively, invoke {@link #installMockForType(Object, TypeLiteral, Options, Annotation...)}.
+ * <p>
  * Mocks installed in {@link org.junit.jupiter.api.BeforeAll} will be present for every test,
  * while mocks installed within a test are cleared after the test has run. Note that you will
  * likely need to use {@link TestInstance.Lifecycle#PER_CLASS} to have a non-static before all method
  * that can access injected beans.
- *
- * Note that as the bean is replaced globally you cannot use parallel test execution, as this will
+ * <p>
+ * Note that as the bean is replaced globally, you cannot use parallel test execution, as this will
  * result in race conditions where mocks from one test are active in another.
- *
  */
 public class QuarkusMock {
 
     /**
-     * Installs a mock for a CDI normal scoped bean
+     * Installs the given {@code mock} for a CDI normal scoped bean identified by given {@code instance}.
+     * Note that the given mock must be assignable to the bean class of the bean; it is <em>not enough</em>
+     * for it to implement only some of the bean types.
      *
      * @param mock The mock object
      * @param instance The CDI normal scoped bean that was injected into your test
@@ -36,56 +40,97 @@ public class QuarkusMock {
      */
     public static <T> void installMockForInstance(T mock, T instance) {
         //mock support does the actual work, but exposes other methods that are not part of the user API
+        checkMockAssignableToInstance(mock, instance);
         MockSupport.installMock(instance, mock, new Options());
     }
 
+    /**
+     * Installs the given {@code mock} for a CDI normal scoped bean identified by given {@code instance}.
+     * Note that the given mock must be assignable to the bean class of the bean; it is <em>not enough</em>
+     * for it to implement only some of the bean types.
+     *
+     * @param mock The mock object
+     * @param instance The CDI normal scoped bean that was injected into your test
+     * @param <T> The bean type
+     */
     public static <T> void installMockForInstance(T mock, T instance, Options options) {
         //mock support does the actual work, but exposes other methods that are not part of the user API
+        checkMockAssignableToInstance(mock, instance);
         MockSupport.installMock(instance, mock, options);
     }
 
     /**
-     * Installs a mock for a CDI normal scoped bean
+     * Installs the given {@code mock} for a CDI normal scoped bean of given {@code type} and with given {@code qualifiers}.
+     * Note that the given mock must be assignable to the bean class of the looked up bean; it is <em>not enough</em>
+     * for it to implement only some of the bean types.
      *
      * @param mock The mock object
-     * @param instance The type of the CDI normal scoped bean to replace
+     * @param type The type of the CDI normal scoped bean to replace
      * @param qualifiers The CDI qualifiers of the bean to mock
      * @param <T> The bean type
      */
-    public static <T> void installMockForType(T mock, Class<? super T> instance, Annotation... qualifiers) {
-        //mock support does the actual work, but exposes other methods that are not part of the user API
-        if (!instance.isAssignableFrom(mock.getClass())) {
-            if (!(instance.getClass().getSuperclass().isAssignableFrom(mock.getClass()))) {
-                throw new RuntimeException(mock
-                        + " is not assignable to type " + instance.getClass().getSuperclass());
-            }
-        }
-        MockSupport.installMock(CDI.current().select(instance, qualifiers).get(), mock, new Options());
+    public static <T> void installMockForType(T mock, Class<? super T> type, Annotation... qualifiers) {
+        Object instance = CDI.current().select(type, qualifiers).get();
+        checkMockAssignableToInstance(mock, instance);
+        MockSupport.installMock(instance, mock, new Options());
     }
 
     /**
-     * Installs a mock for a CDI normal scoped bean by typeLiteral and qualifiers
+     * Installs the given {@code mock} for a CDI normal scoped bean of given {@code type} and with given {@code qualifiers}.
+     * Note that the given mock must be assignable to the bean class of the looked up bean; it is <em>not enough</em>
+     * for it to implement only some of the bean types.
      *
      * @param mock The mock object
-     * @param typeLiteral TypeLiteral representing the required type
+     * @param type The type of the CDI normal scoped bean to replace
      * @param qualifiers The CDI qualifiers of the bean to mock
      * @param <T> The bean type
      */
-    public static <T> void installMockForType(T mock, TypeLiteral<? super T> typeLiteral, Annotation... qualifiers) {
-        MockSupport.installMock(CDI.current().select(typeLiteral, qualifiers).get(), mock, new Options());
+    public static <T> void installMockForType(T mock, TypeLiteral<? super T> type, Annotation... qualifiers) {
+        Object instance = CDI.current().select(type, qualifiers).get();
+        checkMockAssignableToInstance(mock, instance);
+        MockSupport.installMock(instance, mock, new Options());
     }
 
     /**
-     * Installs a mock for a CDI normal scoped bean by typeLiteral and qualifiers
+     * Installs the given {@code mock} for a CDI normal scoped bean of given {@code type} and with given {@code qualifiers}.
+     * Note that the given mock must be assignable to the bean class of the looked up bean; it is <em>not enough</em>
+     * for it to implement only some of the bean types.
      *
      * @param mock The mock object
-     * @param typeLiteral TypeLiteral representing the required type
+     * @param type TypeLiteral representing the required type
      * @param qualifiers The CDI qualifiers of the bean to mock
      * @param <T> The bean type
      */
-    public static <T> void installMockForType(T mock, TypeLiteral<? super T> typeLiteral, Options options,
+    public static <T> void installMockForType(T mock, TypeLiteral<? super T> type, Options options,
             Annotation... qualifiers) {
-        MockSupport.installMock(CDI.current().select(typeLiteral, qualifiers).get(), mock, options);
+        Object instance = CDI.current().select(type, qualifiers).get();
+        checkMockAssignableToInstance(mock, instance);
+        MockSupport.installMock(instance, mock, options);
+    }
+
+    private static <T> void checkMockAssignableToInstance(T mock, T instance) {
+        Class<?> instanceClass;
+        if (instance instanceof MockedThroughWrapper) {
+            instanceClass = instance.getClass();
+            if (instanceClass.getName().endsWith("_ClientProxy")) {
+                // the condition is hideous, but we don't depend on ArC in this module
+                instanceClass = instanceClass.getSuperclass();
+            }
+            // only generated by RestClient Reactive, which always implements exactly 1 interface
+            instanceClass = instanceClass.getInterfaces()[0];
+        } else if (instance instanceof Event<?>) {
+            instanceClass = Event.class;
+        } else if (instance.getClass().getName().endsWith("_ClientProxy")) {
+            // the condition is hideous, but we don't depend on ArC in this module
+            instanceClass = instance.getClass().getSuperclass();
+        } else {
+            throw new RuntimeException("Given instance (" + instance.getClass().getName()
+                    + ") is not a client proxy, only normal scoped beans may be mocked");
+        }
+        Class<?> mockClass = mock.getClass();
+        if (!instanceClass.isAssignableFrom(mockClass)) {
+            throw new RuntimeException("Given mock, which is " + mockClass + ", is not assignable to " + instanceClass);
+        }
     }
 
     /**

